smbclient/_os.py

```diff
@@ -1016,3 +1017,93 @@ def stat(self, follow_symlinks=True):
             if not self._lstat:
                 self._lstat = lstat(self._smb_raw.name)
             return self._lstat
+
+
+def copyfile_server_side(src, dst, **kwargs):
+    """
+    Server side copy of a file
+
+    :param src: The source file.
+    :param dst: The target file.
+    :param kwargs: Common SMB Session arguments for smbclient.
+    """
+
+    if not _is_sameshare(src, dst):
+        raise ValueError(
+            "Server side copy can only occur on the same drive, '%s' must be the same as the dst root '%s'" % (
+                src, dst))
+
+    norm_dst = ntpath.normpath(dst)
+    norm_src = ntpath.normpath(src)
+
+    try:
+        stat(norm_dst, **kwargs)
+    except SMBOSError as err:
+        if err.errno != errno.ENOENT:
+            raise
+    else:
+        raise ValueError("Target %s already exists" % norm_dst)
+
+    with SMBRawIO(norm_src, mode='r', desired_access=FilePipePrinterAccessMask.GENERIC_READ,
+                  create_options=(CreateOptions.FILE_NON_DIRECTORY_FILE), share_access='r',
+                  **kwargs) as raw_src:
+
+        with SMBFileTransaction(raw_src) as transaction_src:
+            ioctl_request(transaction_src, CtlCode.FSCTL_SRV_REQUEST_RESUME_KEY,
+                          flags=IOCTLFlags.SMB2_0_IOCTL_IS_FSCTL,
+                          output_size=32)
+
+        val_resp = SMB2SrvRequestResumeKey()
+        val_resp.unpack(transaction_src.results[0])
+
+        chunks = _get_srv_copy_chunks(transaction_src)
+
+        with SMBRawIO(norm_dst, mode='x', desired_access=FilePipePrinterAccessMask.GENERIC_WRITE,
+                      share_access='r',
+                      create_options=CreateOptions.FILE_NON_DIRECTORY_FILE, **kwargs) as raw_dst:
+
+            with SMBFileTransaction(raw_dst) as transaction_dst:
```

Comment:
This line should be swapped with the for loop. Right now it will send a compound with multiple IOCTL requests which fails as soon as you go over 16 due to the `MAX_COPY_CHUNK_COUNT`. Swapping it means each request only contains up to 16MB of copy chunks hitting the max limit.

Also because of this you will have to move the transaction check below to the proper indentation.

Reply:
swapped :-)